test/common/upstream/load_balancer_fuzz_base.cc

```diff
@@ -0,0 +1,201 @@
+#include "test/common/upstream/load_balancer_fuzz_base.h"
+
+#include "test/common/upstream/utility.h"
+
+namespace Envoy {
+namespace Upstream {
+
+namespace {
+
+constexpr uint32_t MaxNumHostsPerPriorityLevel = 256;
+
+} // namespace
+
+void LoadBalancerFuzzBase::initializeASingleHostSet(
+    const test::common::upstream::SetupPriorityLevel& setup_priority_level,
+    const uint8_t priority_level) {
+  const uint32_t num_hosts_in_priority_level = setup_priority_level.num_hosts_in_priority_level();
+  ENVOY_LOG_MISC(trace, "Will attempt to initialize host set {} with {} hosts.", priority_level,
+                 num_hosts_in_priority_level);
+  MockHostSet& host_set = *priority_set_.getMockHostSet(priority_level);
+  uint32_t hosts_made = 0;
+  // Cap each host set at 256 hosts for efficiency
+  const uint32_t max_num_hosts_in_priority_level = MaxNumHostsPerPriorityLevel;
+  // Leave port clause in for future changes
+  while (hosts_made < std::min(num_hosts_in_priority_level, max_num_hosts_in_priority_level) &&
+         port_ < 65535) {
+    host_set.hosts_.push_back(makeTestHost(info_, "tcp://127.0.0.1:" + std::to_string(port_)));
+    ++port_;
+    ++hosts_made;
+  }
+
+  Fuzz::ProperSubsetSelector subset_selector(setup_priority_level.random_bytestring());
+
+  const std::vector<std::vector<uint8_t>> localities = subset_selector.constructSubsets(
+      {setup_priority_level.num_hosts_locality_one(), setup_priority_level.num_hosts_locality_two(),
+       setup_priority_level.num_hosts_locality_three()},
+      std::min(num_hosts_in_priority_level, max_num_hosts_in_priority_level));
+
+  // Construct three vectors of hosts each representing a locality level, construct
+  // hosts_per_locality from these three vectors
+  HostVector locality_one = {};
+  for (uint8_t index : localities[0]) {
+    ENVOY_LOG_MISC(trace, "Added host at index {} to locality 1", index);
+    locality_one.push_back(host_set.hosts_[index]);
+    locality_indexes_[index] = 0;
+  }
+
+  HostVector locality_two = {};
+  for (uint8_t index : localities[1]) {
+    ENVOY_LOG_MISC(trace, "Added host at index {} to locality 2", index);
+    locality_two.push_back(host_set.hosts_[index]);
+    locality_indexes_[index] = 1;
+  }
+
+  HostVector locality_three = {};
+  for (uint8_t index : localities[2]) {
```

Comment:
it really feels like you can combine all three of these for loops (and below where it's repeated).
Maybe consider a nested for loop over a vector of 3 HostVectors
for (int i = 0; i < 3; i++) {
   // add hosts to each of the three HostVectors.
}

Reply:
Yup, I'll add an array to index just like Adi suggested! Thanks!

Reply:
Switched.